Automattic/jetpack
# Instagram oEmbed: Simplify

Rely on Core's `oEmbed` mechanisms, rather than adding our own embed handler.

Supersedes #11063.
Fixes #12918.
_Possibly_ fixes some other Instagram embed related issues (since this removes our custom embed handler in favor of Core tooling)? Like maybe #16752, #16753, #9331? 

#### Changes proposed in this Pull Request:
- Implement the first (and most crucial) part of pc9Uba-6d-p2: Remove `jetpack_instagram_handler` in favor of Core's own oEmbed system (which, among other things, gives us embed caching for free), while adding
  - a new `jetpack_instagram_oembed_oembed_fetch_url` filter (hooked into `oembed_fetch_url`) to
    - remove any query args from the embed URL (see #17428) and
    - add the now-required auth token for Instagram's oEmbed REST API if present, or redirect the request to WP.com's IG oEmbed proxy instead
  - and a `jetpack_instagram_oembed_remote_get_args` filter (hooked into `oembed_remote_get_args`) to add the required authorization for the WP.com proxy.
- Extract a `build_signed_request` helper method for the latter (in `packages/connection/src/class-client.php`).
- Remove the `instagram_cache_oembed_api_response_body` filter. This allowed returning a boolean value indicating whether to cache or not; if set to true, it would cache for a random amount of time [between one hour and two](https://github.com/Automattic/jetpack/blob/fe6303930772e4b3b276cfba94f651346b6f012d/modules/shortcodes/instagram.php#L185). Core's oEmbeds are cached [for a day](https://github.com/WordPress/WordPress/blob/aef2b539cd665007d3847e27c5e5a3c40afc481c/wp-includes/class-wp-embed.php#L245) by default; caching time can be customized through the `oembed_ttl` filter. _(This is a slight derivation from the original project proposal in `pc9Uba-6d-p2`, but it seemed reasonable to me.)_
- Reenable a number of IG embed unit tests against WP `master`, thus fixing #12918.
- Remove the code that [removes an IG oEmbed handler for an older URL format](https://github.com/Automattic/jetpack/blob/fe6303930772e4b3b276cfba94f651346b6f012d/modules/shortcodes/instagram.php#L87) ([no longer present in Core](https://core.trac.wordpress.org/changeset/44486)), and update a related comment to say that we can also remove the removal of the current IG oEmbed handler [once Core drops it](https://core.trac.wordpress.org/ticket/50861).

#### Jetpack product discussion
pc9Uba-6d-p2

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:

We first test "classic" embeds/shortcodes handling (that doesn't involve Core's oEmbed endpoint that's used by the GB embed block).

- Start a new post, and insert a Classic block.
- Insert the URL of an Instagram post into that block.
- Verify that it expands into the corresponding embed.
- Publish the post, note its ID somewhere, and view it.
- Verify that the embed works there, too.
- To be on the safe side, clear the embed cache by running `yarn docker:wp post meta delete YOUR_POST_ID --all` (using the post ID you noted earlier).
- Reload the post, and verify that the embed still works.

Next, repeat the same without the Classic block by directly pasting the IG post URL into a post. Don't forget to clear oEmbed cache! You should see the embed expand into the embed block, with a preview of the IG post. Once again save, view, clear cache, and reload the post on the frontend.

There's also some good coverage of IG embeds and shortcodes in `tests/php/modules/shortcodes/test-class.instagram.php`, so verify that Travis is green. To run them manually, use

```bash
yarn docker:phpunit --filter=WP_Test_Jetpack_Shortcodes_Instagram
```

#### Proposed changelog entry for your changes:
Instagram embed: Use Core's oEmbed handling, drop custom Jetpack handler.

---

_Big props to @jeherve for his previous work in #11063 and p9GBOq-BC-p2, and to @ebinnion for his work on adding the auth token to requests to IG's oEmbed REST API._
